### PR TITLE
Add playwright e2e tests

### PR DESCRIPTION
**Pull request by nicodh**
Start having e2e tests with playwright based on the browser view

see [README](https://github.com/deltachat/deltachat-desktop/tree/ndh/playwright-e2e-tests/packages/e2e-tests#readme)

### CODE REVIEW

File: packages/e2e-tests/README.md [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+selects "tests" as target for tests and skip the creation of playwright.config.ts
+
+after the install remove the created tests/examples and packages/e2e-tests/tests/example.spec.ts
+
+A convenient alternative to install and use playwright is the VSCode [plugin](https://playwright.dev/docs/getting-started-vscode). It also provides some functionality for running or recording tests.
+
+## Usage
+
+`pnpm e2e --project <chromium | firefox | Chrome>` for headless usage
```
Comment (Simon-Laux) [inline packages/e2e-tests/README.md:19]:
```suggestion
`pnpm -w e2e --project <chromium | firefox | Chrome>` for headless usage
```

-w to run the version on project-root to also include the build

File: packages/e2e-tests/README.md [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+
+cd into packages/e2e-tests
+
+and run `pnpm create playwright`
+
+in the upcoming dialog
+
+selects "tests" as target for tests and skip the creation of playwright.config.ts
+
+after the install remove the created tests/examples and packages/e2e-tests/tests/example.spec.ts
```
Comment (Simon-Laux) [inline packages/e2e-tests/README.md:13]: I doubt that this is a good solution, won't `npx playwright install --with-deps` suffice?

File: packages/target-electron/package.json [hunk 3]
```diff
@@ -40,7 +40,7 @@
     "watch": "pnpm -w --filter=@deltachat-desktop/frontend watch & pnpm watch:compose-frontend",
     "dev": "pnpm build && pnpm start --",
     "dev-inspect": "pnpm build && pnpm start --inspect --",
-    "start": "NODE_OPTIONS=--enable-source-maps electron . --devmode --disable-http-cache --translation-watch",
+    "start": "NODE_OPTIONS=--enable-source-maps electron . --disable-http-cache --translation-watch",
```
Comment (Simon-Laux) [inline packages/target-electron/package.json:43]: why remove `--devmode` here?

Comment (nicodh) [reply packages/target-electron/package.json]: Just an uninteded leftover from a local test

File: packages/e2e-tests/.github/workflows/playwright.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+name: Playwright Tests
```
Comment (WofWca) [inline packages/e2e-tests/.github/workflows/playwright.yml:1]: Does this file do anything? I thought `.github` is supposed to be at the top level of the repo.

**Comment by WofWca** [comment]
Maybe it makes sense to also update other READMEs, e.g. ["How to do end to end testing"](https://github.com/deltachat/deltachat-desktop/blob/21002182831abf909726771d82e65f9cec0f649d/README.md#L38), and the document that it's referencing.

File: packages/e2e-tests/.github/workflows/playwright.yml (hunk 4 above)
Comment (nicodh) [reply packages/e2e-tests/.github/workflows/playwright.yml:1]: It's an auto generated github action that we will integrate in another step later

**Comment by nicodh** [comment]
> Maybe it makes sense to also update other READMEs, e.g. ["How to do end to end testing"](https://github.com/deltachat/deltachat-desktop/blob/21002182831abf909726771d82e65f9cec0f649d/README.md#L38), and the document that it's referencing.

Good point! I updated it

**Comment by nicodh** [comment]
> tests don't start when I also have the browser version open (they should use different default port to fix that I guess)

I added a line in the docs not to start the browser since that happens in the tests

**Comment by WofWca** [comment]
About READMEs again.

- [ ] There is this section, maybe it needs to be updated as well:

https://github.com/deltachat/deltachat-desktop/blob/4d8f77c73e76bdcc157f81d6a68c664aaa61d17a/docs/DEVELOPMENT.md#L104-L110

- [ ] And here is a section about env variables:

https://github.com/deltachat/deltachat-desktop/blob/a5c02433976909fe01a64f3049c96d0d101ae9d8/docs/CLI_FLAGS.md#L22-L29

Also, what about the old tests? This MR removes some of the old testing code, but not all of it (search for `testcafe` and `TEST_DIR`. Is it for later?

**Comment by nicodh** [comment]
> Also, what about the old tests? This MR removes some of the old testing code, but not all of it (search for `testcafe` and `TEST_DIR`. Is it for later?

Good findings! I updated the relevant docs and removed unused test stuff

File: packages/e2e-tests/env.example [hunk 5]
```diff
@@ -0,0 +1 @@
+WEB_PASSWORD=your-password
```
Comment (WofWca) [inline packages/e2e-tests/env.example:1]: Is this file needed here?

**Review by WofWca — CHANGES_REQUESTED** [review]
Tests run for me with `pnpm -w e2e --project chromium`, and with the Playwright extension, in UI mode as well!
Though some warnings are printed and they looks scary, e.g. `src/context.rs:465: can not start io on a context that is not configured`.

The changes to `packages/frontend` (i.e. runtime, production code) are minor ind should have no practical effect, i.e. probably safe to merge and release.

The changes to `packages/target-electron` also seem moderate, except maybe for the `dotenv` dependency that I'm not sure is tree-shaken.

I haven't looked at the actual test code, but it appears to be working! And it doesn't affect production.

Apart from a weird `aria-label` change, I think it's safe to merge for the upcoming production release.

File: packages/frontend/src/components/Dialog/HeaderButton.tsx [hunk 6]
```diff
@@ -13,6 +13,7 @@ type Props = ButtonHTMLAttributes<HTMLButtonElement> & {
   iconSize: number
   rotation?: number
   'aria-label': string
+  dataTestid?: string
```
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/frontend/src/components/Dialog/HeaderButton.tsx:16]: Hmm, how is it included in ButtonHTMLAttributes which comes from react?
If I remove it in the type defintion typescript complains about an unknown prop testId...?

File: packages/target-browser/.env.example [hunk 7]
```diff
@@ -0,0 +1,7 @@
+# this sets the password you have to enter
+# to login in the browser
+WEB_PASSWORD=
+
+# optional
+# with NODE_ENV=test the login is skipped
```
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/target-browser/.env.example]: I don't think its worth. The browser is only for development anyway

File: packages/target-electron/src/application-config.ts [hunk 8]
```diff
@@ -1,11 +1,17 @@
 // @ts-ignore
 import applicationConfig from 'application-config'
+if (process.env.NODE_ENV !== 'production') {
+  // eslint-disable-next-line @typescript-eslint/no-var-requires
+  const { config } = require('dotenv')
```
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/target-electron/src/application-config.ts:5]: its in devDependencies anyway, so it won't be installed in production

File: packages/target-browser/.env.example (hunk 7 above)
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/target-browser/.env.example]: I changed the comment from the frist proposal

File: packages/e2e-tests/package.json [hunk 9]
```diff
@@ -0,0 +1,14 @@
+{
+  "private": true,
+  "name": "@deltachat-desktop/e2e-tests",
+  "type": "module",
+  "license": "GPL-3.0-or-later",
+  "version": "1.46.2-monorepo",
```
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/e2e-tests/package.json]: I updated the version. The license was in there before. Maybe discuss it with Simon...

File: docs/E2E-TESTING.md [hunk 10]
```diff
@@ -1,7 +1,35 @@
...
-## CSS classes used to make testing easier
+## Install playwright
 
-`.test-selector-confirm` - for making confirmation/ok buttons so the test finds them
+after running `pnpm install`
 
-`.test-*` - to mark special elements so the test can find them.
+cd into packages/e2e-tests
+
+and run `npx playwright install --with-deps`
```
[the remark this answers is not in the document]
Comment (nicodh) [reply docs/E2E-TESTING.md:9]: It will only load playwright and execute it. No idea if it has impact, I don't think so

File: packages/frontend/src/components/Dialog/HeaderButton.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (Simon-Laux) [reply packages/frontend/src/components/Dialog/HeaderButton.tsx:16]: you can probably extend that ButtonHTMLAttributes with the additional attribute in global.d.ts, similar how we do it with our `window` extensions.

[the remark this answers is not in the document]
Comment (Simon-Laux) [reply packages/frontend/src/components/Dialog/HeaderButton.tsx:16]: but we can also do that in a follow up pr

File: packages/target-electron/src/application-config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Simon-Laux) [reply packages/target-electron/src/application-config.ts:5]: > its in devDependencies anyway, so it won't be installed in production

that's not how it works here. esbuild bundles it. but I wonder more why require is even allowed here because this is an esm module, I thought there was some workaround to required to get the original `require`

so there is still the tree shake questions remaining, is this still in the bundle in production builds? afaik only the frontend has `NODE_ENV=production` with `pnpm run build4production` or have I missed something.

I would just make this dependent on `--devmode` cli option unless we really have a good workflow that excludes it in production.

[the remark this answers is not in the document]
Comment (Simon-Laux) [reply packages/target-electron/src/application-config.ts:5]: we can convert this into an issue and merge this already for the test release

[the remark this answers is not in the document]
Comment (nicodh) [reply packages/target-electron/src/application-config.ts:5]: I just tried await import('dotenv') instead and it seems to work: found that here: https://dmitripavlutin.com/ecmascript-modules-dynamic-import/

I will test again in production mode

File: packages/frontend/src/components/Dialog/HeaderButton.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (WofWca) [reply packages/frontend/src/components/Dialog/HeaderButton.tsx:16]: Make sure to use kebab-case, i.e.

```typescript
<Button
  data-testid='test'
>
```

for me it doesn't complain this way.

File: packages/target-browser/.env.example (hunk 7 above)
[the remark this answers is not in the document]
Comment (WofWca) [reply packages/target-browser/.env.example]: > The browser is only for development anyway

IDK IDK. From @Simon-Laux 's comments here and there I'm under the impression that it's something that is maybe gonna be in production some day.

File: packages/target-electron/src/application-config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/target-electron/src/application-config.ts]: It works as expected. Even if I uninstall dotenv I can build for production. Added a try catch to make sure it won't influence prod build

**Review by WofWca — APPROVED** [review]

File: packages/target-browser/.env.example (hunk 7 above)
[the remark this answers is not in the document]
Comment (Simon-Laux) [reply packages/target-browser/.env.example]: if someone makes it production they can still make it more secure then. 🤔

**Review by Simon-Laux — APPROVED** [review]

File: packages/frontend/src/components/Dialog/HeaderButton.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (nicodh) [reply packages/frontend/src/components/Dialog/HeaderButton.tsx]: It works anywhere where the {...props} attribute is passed down to the last rendered element, which is the case for Buttons. So you were right and I removed the testId props for buttons
